### PR TITLE
feat: implement dump tool for bc-fusion

### PR DESCRIPTION
**Pull request by j75689**

### CODE REVIEW

File: cmd/dump/main.go [hunk 1]
```diff
@@ -32,75 +35,234 @@ const (
...
+
+func (node *leafNode) Print() string {
+	buf, _ := node.Serialize()
+
+	return "0x" + common.Bytes2Hex(crypto.Keccak256(buf))
+}
+
 // ExportAccounts exports blockchain world state to json.
-func ExportAccounts(app *app.BNBBeaconChain) (appState json.RawMessage, err error) {
+func ExportAccounts(app *app.BNBBeaconChain, outputPath string) (err error) {
```
Comment (zjubfd) [inline cmd/dump/main.go:65]: `ExportAccounts`  ==> `ExportAccountsBalanceWithProof`

File: cmd/dump/main.go [hunk 2]
```diff
@@ -32,75 +35,234 @@ const (
...
+			asset, exist := assets[coin.Denom]
+			if exist {
+				asset.Amount += coin.Amount
+			} else {
+				token, err := app.TokenMapper.GetToken(ctx, coin.Denom)
+				if err != nil {
+					trace(err)
+					return true
+				}
+				assets[coin.Denom] = &types.ExportedAsset{
```
Comment (zjubfd) [inline cmd/dump/main.go:93]: Do we need this `assets` data? 
And we have the summary data about `assets`, dont have to add up from the balance of each account...

File: cmd/dump/main.go [hunk 3]
```diff
@@ -32,75 +35,234 @@ const (
...
 		}
+
 		account := types.ExportedAccount{
 			Address:       addr,
-			AccountNumber: acc.GetAccountNumber(),
-			Coins:         acc.GetCoins(),
+			AccountNumber: namedAcc.GetAccountNumber(),
+			Coins:         allCoins.Sort(),
+			FreeCoins:     coins.Sort(),
+			FrozenCoins:   frozenCoins.Sort(),
```
Comment (zjubfd) [inline cmd/dump/main.go:105]: To save memory, do we need `FreeCoins` and `LockedCoins` and `FreeCoins` here?

File: cmd/dump/main.go [hunk 4]
```diff
@@ -32,75 +35,234 @@ const (
...
+			LockedCoins:   lockedCoins.Sort(),
 		}
 		accounts = append(accounts, &account)
-		mtData = append(mtData, &account)
 
-		if err != nil {
-			fmt.Println(err)
-			return true
+		for index := range allCoins {
+			if allCoins[index].Amount > 0 {
```
Comment (zjubfd) [inline cmd/dump/main.go:111]: only the token that enables cross-chain transfer needs to be added to the tree.

File: cmd/dump/main.go [hunk 5]
```diff
@@ -32,75 +35,234 @@ const (
...
-	exportedProof := make(types.ExportedProofs, len(proofs))
+	maxProofLength := 0
+	exportedProof := make([]*types.ExportedProof, 0, len(proofs))
+	trace("proofs length", len(proofs))
 	for i := 0; i < len(mtData); i++ {
 		proof := proofs[i]
 		nProof := make([]string, 0, len(proof.Siblings))
-		for i := 0; i < len(proof.Siblings); i++ {
-			nProof = append(nProof, "0x"+common.Bytes2Hex(proof.Siblings[i]))
+		for j := 0; j < len(proof.Siblings); j++ {
```
Comment (zjubfd) [inline cmd/dump/main.go:149]: the `path` of the `Siblings` is never used here? `path` variable indicating whether the neighbor is on the left or right.

File: cmd/dump/main.go (hunk 1 above)
Comment (j75689) [reply cmd/dump/main.go]: fixed

File: cmd/dump/main.go (hunk 2 above)
Comment (j75689) [reply cmd/dump/main.go]: yeah, we can remove it

File: cmd/dump/main.go (hunk 3 above)
Comment (j75689) [reply cmd/dump/main.go]: yeah, we can remove them

File: cmd/dump/main.go (hunk 4 above)
Comment (j75689) [reply cmd/dump/main.go:104]: fixed

File: cmd/dump/main.go (hunk 5 above)
Comment (j75689) [reply cmd/dump/main.go:142]: the hash will be compared before calculate

reference:
https://github.com/OpenZeppelin/openzeppelin-contracts/blob/88211e8fbad2e690e62c0dbfc4b14b2d2b6ab566/contracts/utils/cryptography/MerkleProof.sol#L217
https://github.com/txaty/go-merkletree/blob/ccfc3d24289d9c28111edc02f688b8bef9f45fb7/merkle_tree.go#L124
https://github.com/txaty/go-merkletree/blob/ccfc3d24289d9c28111edc02f688b8bef9f45fb7/proof_gen.go#L82

**Pull request merged**